File: DotNetOnion/Cells/CellCreatedFast.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using DotNetOnion.Helpers;

namespace DotNetOnion.Cells
{
    public class CellCreatedFast : Cell
    {
        public override byte Command => 0x06;

        public byte[] Y { get; set; }
        public byte[] DerivativeKeyData { get; set; }

        public override void Deserialize(BinaryReader reader)
        {
            Y = reader.ReadBytes(Constants.HashLength);
            DerivativeKeyData = reader.ReadBytes(Constants.HashLength);
        }

        public override void Serialize(BinaryWriter writer)
        {
            writer.Write(Y);
            writer.Write(DerivativeKeyData); 
        }
    }
}

File: NOnion/Cells/CellCreatedFast.fs
namespace NOnion.Cells

open System.IO

open NOnion

type CellCreatedFast =
    private
        {
            Y: array<byte>
            DerivativeKeyData: array<byte>
        }

    static member Deserialize(reader: BinaryReader) =
        {
            Y = reader.ReadBytes Constants.HashLength
            DerivativeKeyData = reader.ReadBytes Constants.HashLength
        }
        :> ICell

    interface ICell with

        member __.Command = 6uy

        member self.Serialize writer =
            writer.Write self.Y
            writer.Write self.DerivativeKeyData

    interface ICreatedCell with
        member self.ServerHandshake = self.Y

        member self.DerivativeKey = self.DerivativeKeyData
